src/dataflow/src/render/upsert.rs

```diff
@@ -0,0 +1,271 @@
+                    {
+                        let value = values
+                            .entry(cap.delayed(&time))
+                            .or_insert_with(HashMap::new)
+                            .entry(key)
+                            .or_insert_with(Default::default);
+
+                        if let Some(new_offset) = position {
+                            if let Some(offset) = value.1 {
+                                if offset < new_offset {
+                                    *value = (val, position);
+                                }
+                            } else {
+                                *value = (val, position);
+                            }
+                        }
+                    }
+                });
+
+                // Produce (key, val) pairs at any complete times.
+                for (cap, map) in values.iter_mut() {
+                    if !input.frontier.less_equal(cap.time()) {
+                        let mut session = output.session(cap);
+                        for (key, val) in map.drain() {
+                            session.give(((key, val), cap.time().clone()))
+                        }
+                    }
+                }
+                // Discard entries, capabilities for complete times.
+                values.retain(|_cap, map| !map.is_empty());
+            }
+        },
+    )
+}
+
+/// Apply a filter followed by a project to an upsert stream.
+fn apply_linear_operators<G>(
+    stream: &Stream<G, (Row, Option<Row>, Timestamp)>,
+    operator: &LinearOperator,
+    src_type: &RelationType,
+) -> (
+    Stream<G, (Row, Option<Row>, Timestamp)>,
+    Stream<G, DataflowError>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    // Determine replacement values for unused columns.
+    // This is copied over from applying linear operators to
+    // the non-upsert case.
+    // At this point, the stream consists of (key, key+payload, time)
+    // so it is ok to delete replace unused values in key+payload
+    // but not to replace unused values in key because that would cause
+    // errors in arrange_from_upsert
+    let position_or = (0..src_type.arity())
+        .map(|col| {
+            if operator.projection.contains(&col) {
+                Some(col)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // If a row does not match a predicate whose support lies in the key
+    // columns, it can be tossed away entirely. If a row does not match
+    // a predicate whose support is not contained in the key columns, it
+    // must be replaced with (key, None) in case if there was previously a
+    // row with the same key that matched the predicate.
+    let key_col_len = src_type.keys[0].len();
+    let mut predicates = operator.predicates.clone();
+    let mut key_predicates = Vec::new();
+    predicates.retain(|p| {
+        let key_predicate = p.support().into_iter().all(|c| c < key_col_len);
```

Comment:
Is it right that the key columns are always the first columns? It may be, but I didn't realize that!

Reply:
In `decode_upsert`, after the decoding happens, there is a map that prepends the key columns to the value columns that way the resulting arrangement produced by `arrange_from_upsert` follows the convention that all other arrangements are in, namely (key, <all columns including key columns>). 

But now that you point this out, I think I should pull the map out of `decode_upsert` to `upsert.rs` so we have greater visibility into when the prepending of key columns to value columns happen (and perhaps consider shifting the prepending to occur later in the process)?